ros2/inference_base.py

```diff
@@ -0,0 +1,231 @@
+import threading
+from abc import ABC, abstractmethod
+from typing import List, Union, Dict, Any
+
+import rclpy
+from rclpy.node import Node
+from std_msgs.msg import Float32
+from sensor_msgs.msg import Image
+
+# Protobuf generated modules
+from config.proto import config_pb2
+from config.proto import ai_pb2
+
+
+class InferenceBase(Node, ABC):
+    """Abstract base class for inference nodes.
+    
+    This class provides the common infrastructure for inference nodes including:
+    - ROS2 node initialization
+    - Publisher/subscriber setup
+    - State tracking for both camera and encoder data
+    - Thread-safe message handling
+    
+    Subclasses must implement the inference logic by overriding abstract methods.
+    """
+    
+    def __init__(self, node_name: str, node_id: int, config: config_pb2.Config):
+        super().__init__(node_name)
+        
+        self.node_id = node_id
+        self.config = config
+        
+        # Validate operation mode (subclasses can override this check)
+        if not self._is_valid_operation_mode(config.general.operation_mode):
+            self.get_logger().error(f"Invalid operation mode for {self.__class__.__name__}")
+            return
+        
+        # Initialize publishers
+        self.action_publishers: List[rclpy.publisher.Publisher] = []
+        for i in range(len(config.ai.publish_topics)):
+            topic = config.ai.publish_topics[i]
+            self.action_publishers.append(self.create_publisher(Float32, topic, 10))
+        
+        # Separate camera and encoder topics for better organization
+        self.camera_topics = []
+        self.encoder_topics = []
```

Comment:
maybe we can create `subscription_topics`, and `publish_topics` and fill them in subclass by instantiating camera_topic, encoder_topic and so on.